### PR TITLE
Fix to make `JsonObject` type partial

### PR DESCRIPTION
**Pull request by nvie**
This fixes the `JsonObject` type to recognize that keys could be validly `undefined`.

Wrong behavior:

```tsx
const obj: JsonObject = ...;
obj.whatever;   // => Json ❌
```

New/correct behavior:

```tsx
const obj: JsonObject = ...;
obj.whatever;   // => Json | undefined ✅
```

### CODE REVIEW

**Comment by GuillaumeSalles** [comment]
My initial thought was that we were supposed to do the opposite, but maybe I'm wrong 🤔 

From what I understand, `undefined` is not serializable to json, so we shouldn't be able to set have undefined values on a `LiveObject`. `UpdateObjectOp.data` should be a `JsonObject` or a `Record<string, Json>` instead of a `Partial<JsonObject>`? Sorry about the confusion, I'm still not 100% of what's the correct type

**Comment by GuillaumeSalles** [comment]
My initial thought was that we were supposed to do the opposite, but maybe I'm wrong 🤔 

From what I understand, `undefined` is not serializable to json, so we shouldn't be able to set have undefined values on a `LiveObject`. `UpdateObjectOp.data` should be a `JsonObject` or a `Record<string, Json>` instead of a `Partial<JsonObject>`? Sorry about the confusion, I'm still not 100% of what's the correct type

**Comment by nvie** [comment]
Yeah, its been confusing for sure. My wrong thinking before comes from the fact that `undefined` does not have a JSON _representation_, but you can still work with them in TypeScript land on the input and output side. I hope the following examples help to illustrate that.

On the serializing side, not allowing `undefined` in inputs was too restrictive. It’s okay to send in `{ a: 1, b: undefined }`, which is the same as sending in `{ a: 1 }`, because they both produce the same string when using `JSON.stringify()`.

On the parsing side, `{ [key: string]: Json }` means that all potential thinkable keys on an object are always going to be `Json` in TypeScript. They almost never are! For example: `JSON.parse("{ a: 1 }").someRandomProperty` is definitely going to be `undefined`.

#### The old/wrong way

```ts
const x: JsonObject = { a: 1, b: undefined };
//                            ^ ❌ Errors here
JSON.stringify(x);  // Even though this call is fine!
```

Also:

```ts
const y: JsonObject = JSON.parse("{ a: 1 }");
const a: Json = y.a;  // Correct
const b: Json = y.b;  // ❌ Incorrect! It’s actually `undefined`, but TypeScript misses that truth
```

#### The new/correct way

```ts
// ✅ Fine now
const x: JsonObject = { a: 1, b: undefined };
JSON.stringify(x);
```

Also:

```ts
const y: JsonObject = JSON.parse("{ a: 1 }");
const a = y.a;
const b = y.b;
//    ^ ✅ Now correctly infers that these could be Json | undefined
```

**Comment by nvie** [comment]
tl;dr `undefined` itself isn’t a legal `Json` value, but a `JsonObject` can still have implicitly-missing or explicitly-missing (through `undefined`) entries.

**Comment by nvie** [comment]
Tying this back to your `UpdateObjectOp.data` use case: I think from that perspective, we should simply ignore any keys with `undefined` values ourselves (skipping over them when looping). Because depending on where the JSON value is generated, it’s possible that there are explicit-undefined keys (and thus more entries to loop over), which should be treated as implicit-undefined keys (which would not pop up when looping all entries).

I hate the fact that implicit-missing and explicit-missing values are observable (e.g. when looping over entries), but this is the most honest reflection of the possible realities that are the consequence of JS’ language design.

**Comment by GuillaumeSalles** [comment]
Thank you for the detailed explanation! I now understand why you implemented the fix that way. The new `JsonObject` type is a better representation of a javascript object created by a `JSON.parse`.

> I hate the fact that implicit-missing and explicit-missing values are observable (e.g. when looping over entries), but this is the most honest reflection of the possible realities that are the consequence of JS’ language design.

Yep, we're on the same page. "The looping over entries" is what I was thinking about when I mentioned serialization. My intuition is that if we allow `LiveObject<{ x?: number }>`, we might give the impression that the LiveObject keys will be the same on all clients after `liveObject.set("x", undefined)`. But at the same time, it could help us refine the type of `LiveObject.delete` by allowing only keys where the value can be `undefined`?

**Review by GuillaumeSalles — APPROVED** [review]

**Comment by nvie** [comment]
> My intuition is that if we allow `LiveObject<{ x?: number }>`, we might give the impression that the LiveObject keys will be the same on all clients after `liveObject.set("x", undefined)`. But at the same time, it could help us refine the type of `LiveObject.delete` by allowing only keys where the value can be `undefined`?

Ohh... great point! I didn't even consider that use case yet, but I think that this now indeed naturally solves for that. In `LiveObject.set(x, y)`, `y` should never be undefined (or better… it should implicitly translate to `LiveObject.delete(x)` instead), and indeed: not allowing `.delete()` on keys that aren't declared to be optional would be an amazing DX if you ask me :)
